src/IssueViz/Editor/SelectedIssueTagging/Adornment/IssueLocationAdornmentTagger.cs

```diff
@@ -40,11 +66,85 @@ protected override TagSpan<IntraTextAdornmentTag> CreateTagSpan(ISelectedIssueLo
             // should have zero length spans. Overriding this method allows control
             // over the tag spans.
             var adornmentSpan = new SnapshotSpan(trackedTag.Location.Span.Value.Start, 0);
-            var adornment = new IssueLocationAdornment(trackedTag, wpfView.FormattedLineSource);
+            var adornment = adornmentCache.CreateOrUpdate(trackedTag.Location);
 
-            // If we don't call Measure here the tag is positioned incorrectly
-            adornment.Measure(new System.Windows.Size(double.PositiveInfinity, double.PositiveInfinity));
             return new TagSpan<IntraTextAdornmentTag>(adornmentSpan, new IntraTextAdornmentTag(adornment, null, PositionAffinity.Predecessor));
         }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                tagAggregator.BatchedTagsChanged -= OnTagsChanged;
+            }
+            base.Dispose(disposing);
+        }
+
+        #region Adornment cache
+
+        internal interface ICachingAdornmentFactory
```

Comment:
Maybe extract this to another file? I think it has enough in it that it deserves its own file.